src/main.rs

```diff
@@ -0,0 +1,223 @@
+		Some("devpts"), nix::mount::MsFlags::empty(), None::<&str>
+	).expect("failed to mount /dev/pts");
+
+	nix::mount::mount(None::<&str>, &concat_absolute(&cfg.rootfs, "/dev/shm"),
+		Some("tmpfs"), nix::mount::MsFlags::empty(), None::<&str>
+	).expect("failed to mount /dev/shm");
+
+	nix::mount::mount(None::<&str>, &concat_absolute(&cfg.rootfs, "/run"),
+		Some("tmpfs"), nix::mount::MsFlags::empty(), None::<&str>
+	).expect("failed to mount /run");
+
+	nix::mount::mount(None::<&str>, &concat_absolute(&cfg.rootfs, "/tmp"),
+		Some("tmpfs"), nix::mount::MsFlags::empty(), None::<&str>
+	).expect("failed to mount /tmp");
+
+	// Perform bind mounts requested by user.
+	for bm in &cfg.bind_mounts {
+		nix::mount::mount(Some(&bm.src), &concat_absolute(&cfg.rootfs, &bm.dest),
+			None::<&str>, nix::mount::MsFlags::MS_BIND | nix::mount::MsFlags::MS_REC, None::<&str>
+		).expect("failed to perform bind mount");
+	}
+
+	// chroot() and change the current directory to /.
+	nix::unistd::chroot(&cfg.rootfs)
+		.expect("failed to chroot()");
+	nix::unistd::chdir("/")
+		.expect("failed to chdir() to root directory");
+
+	// TODO: We could drop privileges here.
+	//       (However, cbuildrt does not really protect against malicious sandbox escapes.)
+
+	// fork() and execve() in the child.
+	// (We cannot use Rust's high-level API since we need to reap orphans.)
+	let fork_result = unsafe { nix::unistd::fork() };
+	if let Ok(nix::unistd::ForkResult::Child) = fork_result {
+		let exec_result = nix::unistd::execvp(
+			&CString::new(cfg.args[0].as_str()).unwrap(),
+			&cfg.args.iter().map(|a| CString::new(a.as_str()).unwrap())
+				.collect::<Vec<_>>()
+		);
+		println!("error when executing program: {}", exec_result.unwrap_err());
+		exit(1);
+	}
+
+	let child_pid = match fork_result {
+		Ok(nix::unistd::ForkResult::Parent{child}) => child,
+		_ => panic!("failed to fork from init"),
+	};
+
+	loop {
+		// Now, let's wait for the child to terminate.
+		let child_status = nix::sys::wait::wait()
+			.expect("failed to wait for children");
+		if let nix::sys::wait::WaitStatus::Exited(pid, code) = child_status {
+			if pid == child_pid {
+				if code != 0 {
+					println!("child returned non-zero exit code");
+				}
+				exit(code);
+			}
+		}
+	}
+}
+
+fn main() {
+	let cfg = make_config_from_cli();
+
+	let euid = nix::unistd::geteuid();
+	let egid = nix::unistd::getegid();
+
+	// Enter the user namespace and let children enter a new PID namespace.
+	// We cannot do mounts in this process yet, as this the process itself
+	// is not moved to the new PID namespace.
+	nix::sched::unshare(
+		nix::sched::CloneFlags::CLONE_NEWUSER
+		| nix::sched::CloneFlags::CLONE_NEWPID
+	).expect("failed to unshare()");
+
+	// Write the uid_map and gid_map files.
+	let mut setgroups_f = File::create("/proc/self/setgroups")
+		.expect("unable to open setgroups");
+	setgroups_f.write_all("deny".as_bytes())
+		.expect("unable to write setgroups file");
```

Comment:
You could use [`std::fs::write`](https://doc.rust-lang.org/std/fs/fn.write.html) here instead